meson_options.txt

```diff
@@ -197,12 +197,12 @@ option('newlib-retargetable-locking', type: 'boolean', value: true,
 #
 option('thread-local-storage', type: 'combo', choices: ['true', 'false', 'auto'], value: 'auto',
        description: 'use thread local storage for static data (default: automatically detect support)')
-option('tls-model', type: 'string', value: 'local-exec',
+option('tls-model', type: 'combo', choices: ['global-dynamic', 'local-dynamic', 'initial-exec', 'local-exec'], value: 'local-exec',
```

Comment:
This should probably add a 'none' option to allow configurations which don't want to specify -ftls-model to the compiler?

Reply:
>     * **thread-local-storage**: This sets `NEWLIB_TLS` and `PICOLIBC_TLS`

Yeah, probably don't need both defines. Let's keep PICOLIBC_TLS

>     * **NEWLIB_TLS**: define `NEWLIB_THREAD_LOCAL`.
> 
>     * **PICOLIBC_TLS**: guard for `_set_tls` et al. at the call sites and the header[*1].
> 
>     * **HAVE_PICOLIBC_TLS_API**: also guard `_set_tls` (in `test/tls.c`) activated by `has_picolibc_tls_api`

This gets set when there is machine-specific code for the _set_tls function. Picolibc has support for arm, aarch64 and risc-v.

>     * **has_picolibc_tls_api**: set by `newlib/libc/picolib/meson.build` when given arch has `machine/*/tls.c`, which has `_set_tls()` body, with a comment saying _"This code is duplicated in ..."_[*2].

right, the _set_tls function is architecture dependent, and has to agree with the tls-model in use with the C compiler.

> I thought `thread_local_storage` should be tri-state. And take `tls-model` if and only if `thread_local_storage` is true or enabled. But, i686 and amd64 is building with `-Dthread-local-storage=false`. Why?

because picolibc doesn't have support for _set_tls on i686 and amd64, so you can't use thread local storage on that platform. Adding _set_tls for i686 or amd64 would be possible as the C compiler clearly supports it, but it's a bit complicated and I haven't ever bothered as neither of these are really viable for embedded systems, and if all you're doing is testing picolibc on them, you probably don't need TLS.

> [*1]: Do you like to protect function declarations like we do in `picolibc/newlib/libc/include/picotls.h`? I don't think we need to protect them. `picotls.h` has inline function `_tls_size()` and we might want to guard it, though. [*2]: In the case of RISC-V, the comment is wrong. There is no `picocrt/machine/riscv/crt0.c` as commented.

This header file is exposed to applications, so I think it's nice to have a simple compile error when you attempt to use TLS features and the library isn't built to support them.